src/simulators/circuit_executor.hpp

```diff
@@ -0,0 +1,1174 @@
+
+  // insert runtime noise sample ops here
+  RngEngine rng;
+  rng.set_seed(circ.seed);
+
+  int_t par_shots = (int_t)get_max_parallel_shots(circ, noise);
+  par_shots = std::min((int_t)parallel_shots_, par_shots);
+  std::vector<ExperimentResult> par_results(par_shots);
+
+  uint_t num_shots = circ.shots;
+  uint_t seed_begin = circ.seed;
+
+  // MPI distribution settings
+  std::vector<ClassicalRegister> cregs;
+  reg_t shot_begin(distributed_procs_);
+  reg_t shot_end(distributed_procs_);
+  for (int_t i = 0; i < distributed_procs_; i++) {
+    shot_begin[i] = circ.shots * i / distributed_procs_;
+    shot_end[i] = circ.shots * (i + 1) / distributed_procs_;
+  }
+  num_shots = shot_end[distributed_rank_] - shot_begin[distributed_rank_];
+  seed_begin += shot_begin[distributed_rank_];
+  cregs.resize(circ.shots);
+
+  // run each shot
+  auto run_circuit_lambda = [this, &par_results, circ, noise, config, par_shots,
+                             sample_noise, num_shots, seed_begin, shot_begin,
+                             &cregs](int_t i) {
+    Noise::NoiseModel dummy_noise;
+    state_t state;
+    uint_t i_shot, shot_end;
+    i_shot = num_shots * i / par_shots;
+    shot_end = num_shots * (i + 1) / par_shots;
+
+    auto fusion_pass = transpile_fusion(circ.opset(), config);
+
+    // Set state config
+    state.set_config(config);
+    state.set_parallelization(this->parallel_state_update_);
+    state.set_global_phase(circ.global_phase_angle);
+    state.enable_density_matrix(!has_statevector_ops_);
+
+    state.set_distribution(this->num_process_per_experiment_);
+    state.set_num_global_qubits(circ.num_qubits);
+
+    for (; i_shot < shot_end; i_shot++) {
+      RngEngine rng;
+      rng.set_seed(seed_begin + i_shot);
+
+      int max_matrix_qubits;
+      Circuit circ_opt;
+      if (sample_noise)
+        circ_opt = noise.sample_noise(circ, rng);
+      else
+        circ_opt = circ;
+      fusion_pass.optimize_circuit(circ_opt, dummy_noise, state.opset(),
+                                   par_results[i]);
+      max_matrix_qubits = get_max_matrix_qubits(circ_opt);
+      state.set_max_matrix_qubits(max_matrix_qubits);
+      state.enable_cuStateVec(cuStateVec_enable_);
+
+      state.allocate(circ.num_qubits, circ.num_qubits);
+      state.initialize_qreg(circ.num_qubits);
+      state.initialize_creg(circ.num_memory, circ.num_registers);
+
+      state.apply_ops(circ_opt.ops.cbegin(), circ_opt.ops.cend(),
+                      par_results[i], rng, true);
+      if (distributed_procs_ > 1) {
+        // save creg to be gathered
+        cregs[shot_begin[distributed_rank_] + i_shot] = state.creg();
+      } else {
+        par_results[i].save_count_data(state.creg(), save_creg_memory_);
+      }
+    }
+    state.add_metadata(par_results[i]);
+  };
+  Utils::apply_omp_parallel_for((par_shots > 1), 0, par_shots,
+                                run_circuit_lambda);
+
+  // gather cregs on MPI processes and save to result
```

Comment:
The below can be in `#AER_MPI` (`gather_creg_memory ` is also), I guess.

Reply:
`#ifdef AER_MPI` inserted